### PR TITLE
Document the hash table runtime stats

### PR DESCRIPTION
**Pull request by duanmeng**

### CODE REVIEW

File: velox/docs/monitoring/stats.rst [hunk 1]
```diff
@@ -39,3 +39,25 @@ Memory Arbitration
...
+Task Execution
+------------------
+.. list-table::
+   :widths: 40 10 50
+   :header-rows: 1
+
+   * - Stats Name
+     - Type
+     - Description
+   * - hashtable.capacity
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:52]: The stats above are reported for any kind of operator. However, these stats are reported only by HashBuild and HashAggregation operators. 

Let's figure out how to clarify this in the docs. One option might be to create a section per operator.

File: velox/exec/HashBuild.cpp [hunk 2]
```diff
@@ -45,6 +45,12 @@ BlockingReason fromStateToBlockingReason(HashBuild::State state) {
       VELOX_UNREACHABLE(HashBuild::stateName(state));
   }
 }
+
+const std::string kCapacity{"hashtable.capacity"};
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/exec/HashBuild.cpp:49]: Should we put them in the `HashTable.h` and use `constexpr std::string_view`?

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/exec/HashBuild.cpp:49]: Sounds reasonable.

File: velox/docs/monitoring/stats.rst (hunk 1 above)
Comment (duanmeng) [reply velox/docs/monitoring/stats.rst:52]: Add a `HashTable` section as they are used in both HashBuild and HashAggregation operators and `HashTable` is a very critical component.

File: velox/docs/monitoring/stats.rst [hunk 3]
```diff
@@ -39,3 +39,25 @@ Memory Arbitration
      - kByte
      - The reclaimed memory bytes of an operator during the memory arbitration.
        This stats only applies for spillable operators.
+
+HashTable
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:43]: Let's state explicitly which operators are expected to report these stats.

File: velox/exec/HashBuild.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/exec/HashBuild.cpp]: updated.

File: velox/exec/HashTable.h [hunk 4]
```diff
@@ -23,6 +23,11 @@
 
 namespace facebook::velox::exec {
 
+constexpr std::string_view kCapacity{"hashtable.capacity"};
```
Comment (mbasmanova) [inline velox/exec/HashTable.h:26]: Please, add comments to explain that these are the names of the runtime stats reported by operators that use the HashTable.

Comment (mbasmanova) [inline velox/exec/HashTable.h:26]: > kCapacity

These names cannot appear as is in velox::exec scope. There is simply not enough context in these: velox::exec::kCapacity.

Perhaps, move these inside the HashTable class?

File: velox/exec/HashAggregation.cpp [hunk 5]
```diff
@@ -180,12 +180,12 @@ void HashAggregation::updateRuntimeStats() {
     }
   }
 
-  runtimeStats["hashtable.capacity"] = RuntimeMetric(hashTableStats.capacity);
-  runtimeStats["hashtable.numRehashes"] =
+  runtimeStats[std::string(kCapacity)] = RuntimeMetric(hashTableStats.capacity);
```
Comment (mbasmanova) [inline velox/exec/HashAggregation.cpp:183]: it is a bit tedious to wrap names in std::string? Can we figure out how to not require that?

```
runtimeStats[kCapacity] = RuntimeMetric(hashTableStats.capacity);
```

File: velox/exec/HashTable.h (hunk 4 above)
Comment (duanmeng) [reply velox/exec/HashTable.h:26]: Sounds good. It would avoid copying the string.

File: velox/exec/HashAggregation.cpp (hunk 5 above)
Comment (duanmeng) [reply velox/exec/HashAggregation.cpp:183]: I will put them into the HashTable class as static members and use std::string.

File: velox/exec/HashTable.h (hunk 4 above)
Comment (duanmeng) [reply velox/exec/HashTable.h]: Fixed

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, I've resolved all the comments according to your guidance, could you please take another look? Thanks a lot.

File: velox/docs/monitoring/stats.rst [hunk 6]
```diff
@@ -39,3 +39,25 @@ Memory Arbitration
      - kByte
      - The reclaimed memory bytes of an operator during the memory arbitration.
        This stats only applies for spillable operators.
+
+HashBuild, HashAggregation
+------------------
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:44]: --- needs to be extended

Please, generate the docs locally and verify there are not errors / warnings and the result looks nice.

File: velox/docs/monitoring/stats.rst [hunk 7]
```diff
@@ -39,3 +39,25 @@ Memory Arbitration
...
+.. list-table::
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:45]: nit: add empty line between the header and this table

File: velox/docs/monitoring/stats.rst [hunk 8]
```diff
@@ -39,3 +39,25 @@ Memory Arbitration
...
+   :widths: 40 10 50
+   :header-rows: 1
+
+   * - Stats Name
+     - Type
+     - Description
+   * - hashtable.capacity
+     - kNone
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:53]: > kNone

What does that mean? Would readers understand this? Let's find a more intuitive way to describe the type of a stat.

File: velox/exec/HashAggregation.cpp [hunk 9]
```diff
@@ -180,12 +180,13 @@ void HashAggregation::updateRuntimeStats() {
     }
   }
 
-  runtimeStats["hashtable.capacity"] = RuntimeMetric(hashTableStats.capacity);
-  runtimeStats["hashtable.numRehashes"] =
+  runtimeStats[std::string(HashTable<true>::kCapacity)] =
```
Comment (mbasmanova) [inline velox/exec/HashAggregation.cpp:183]: > HashTable<true>...

It is a bit weird to use the template parameter here. Maybe move these to HashTableBase which is not a template?

File: velox/exec/HashAggregation.cpp [hunk 10]
```diff
@@ -180,12 +180,13 @@ void HashAggregation::updateRuntimeStats() {
...
+      RuntimeMetric(hashTableStats.capacity);
+  runtimeStats[std::string(HashTable<true>::kNumRehashes)] =
```
Comment (mbasmanova) [inline velox/exec/HashAggregation.cpp:185]: Can we get rid of std::string here?

**Review by mbasmanova — COMMENTED** [review]
@duanmeng Thank you for iterating on this change. Some further comments.

File: velox/docs/monitoring/stats.rst (hunk 8 above)
Comment (duanmeng) [reply velox/docs/monitoring/stats.rst:53]: kNone, kNanos are RuntimeCounter units `enum class Unit { kNone, kNanos, kBytes };`

Shall we use **Count**  or ** Quantity*?

File: velox/exec/HashAggregation.cpp (hunk 9 above)
Comment (duanmeng) [reply velox/exec/HashAggregation.cpp]: Sure, done.

File: velox/exec/HashAggregation.cpp (hunk 10 above)
Comment (duanmeng) [reply velox/exec/HashAggregation.cpp]: Fixed

File: velox/docs/monitoring/stats.rst (hunk 6 above)
Comment (duanmeng) [reply velox/docs/monitoring/stats.rst]: Fixed

Comment (duanmeng) [reply velox/docs/monitoring/stats.rst]: The Clion has a plugin to preview rst files. I checked it after modifying it.

File: velox/docs/monitoring/stats.rst (hunk 7 above)
Comment (duanmeng) [reply velox/docs/monitoring/stats.rst:47]: Fixed

File: velox/docs/monitoring/stats.rst [hunk 11]
```diff
@@ -39,3 +40,26 @@ Memory Arbitration
...
+
+HashBuild, HashAggregation
+--------------------------
+
+.. list-table::
+   :widths: 40 10 50
+   :header-rows: 1
+
+   * - Stats Name
+     - Type
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:52]: I'm thinking that all runtime stats values are integers. Some are simple counters, while others have a unit (bytes or nanoseconds). Hence, perhaps, the table header could say "Unit" and the values will be either `<empty>` (default, simple counter), bytes, or nanoseconds.

What do you think?

```
struct RuntimeMetric {
  // Sum, min, max have the same unit, count has kNone.
  RuntimeCounter::Unit unit;
  int64_t sum{0};
  int64_t count{0};
  int64_t min{std::numeric_limits<int64_t>::max()};
  int64_t max{std::numeric_limits<int64_t>::min()};
```

File: velox/docs/monitoring/stats.rst [hunk 12]
```diff
@@ -39,3 +40,26 @@ Memory Arbitration
...
        This stats only applies for spillable operators.
+
+HashBuild, HashAggregation
+--------------------------
+
+.. list-table::
+   :widths: 40 10 50
+   :header-rows: 1
+
+   * - Stats Name
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:51]: > Stats Name

Perhaps, this can be just "Name".

File: velox/exec/HashTable.cpp [hunk 13]
```diff
@@ -1978,7 +1983,6 @@ void HashTable<ignoreNullKeys>::checkConsistency() const {
       numTombstone,
       numDistinct_);
 }
-
```
Comment (mbasmanova) [inline velox/exec/HashTable.cpp:1981]: unintended change?

File: velox/exec/HashTable.h [hunk 14]
```diff
@@ -22,7 +22,6 @@
 #include "velox/exec/VectorHasher.h"
 
 namespace facebook::velox::exec {
-
```
Comment (mbasmanova) [inline velox/exec/HashTable.h:25]: unintended change?

File: velox/exec/HashTable.h [hunk 15]
```diff
@@ -602,7 +608,6 @@ class HashTable : public BaseHashTable {
   auto& testingOtherTables() const {
     return otherTables_;
   }
-
```
Comment (mbasmanova) [inline velox/exec/HashTable.h:605]: unintended change?

File: velox/docs/monitoring/stats.rst [hunk 16]
```diff
@@ -18,6 +18,7 @@ analysis.
 
 Memory Arbitration
 ------------------
+
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:21]: Let's add a sentence here to explain that these stats are available for all operators.

File: velox/docs/monitoring/stats.rst [hunk 17]
```diff
@@ -39,3 +40,26 @@ Memory Arbitration
      - kByte
      - The reclaimed memory bytes of an operator during the memory arbitration.
        This stats only applies for spillable operators.
+
+HashBuild, HashAggregation
+--------------------------
+
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:46]: Let's clarify somehow that these stats are reported only by HashBuild and HashAggregation operators.

File: velox/docs/monitoring/stats.rst [hunk 18]
```diff
@@ -26,7 +27,7 @@ Memory Arbitration
      - Type
      - Description
    * - memoryReclaimCount
-     - kNone
+     - Quantity
```
Comment (xiaoxmeng) [inline velox/docs/monitoring/stats.rst:30]: Shall we update the code definition as well?

File: velox/docs/monitoring/stats.rst (hunk 11 above)
Comment (duanmeng) [reply velox/docs/monitoring/stats.rst]: Sounds great. Updated.

File: velox/docs/monitoring/stats.rst (hunk 18 above)
Comment (duanmeng) [reply velox/docs/monitoring/stats.rst]: use `<empty>` according to this comment https://github.com/facebookincubator/velox/pull/9255#discussion_r1539723555

**Comment by duanmeng** [comment]
@mbasmanova I've resolved all the comments, could you please help take another look?

File: velox/exec/HashTable.h [hunk 19]
```diff
@@ -123,6 +123,13 @@ class BaseHashTable {
 
   static constexpr int8_t kNoSpillInputStartPartitionBit = -1;
 
+  /// The name of the runtime stats collected and reported by operators that use
+  /// the HashTable (HashBuild, HashAggregation).
+  static const std::string kCapacity;
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.h:128]: Use inline static, then you don't declare their storage in cc.

File: velox/docs/monitoring/stats.rst [hunk 20]
```diff
@@ -18,24 +18,50 @@ analysis.
...
    :widths: 40 10 50
    :header-rows: 1
 
    * - Stats Name
-     - Type
+     - Unit
      - Description
    * - memoryReclaimCount
-     - kNone
+     - <empty>
```
Comment (xiaoxmeng) [inline velox/docs/monitoring/stats.rst:31]: Not sure we name this as kCount type?

**Review by xiaoxmeng — APPROVED** [review]
@duanmeng LGTM. Also have @mbasmanova take a look. Thank you!

File: velox/docs/monitoring/stats.rst (hunk 20 above)
Comment (mbasmanova) [reply velox/docs/monitoring/stats.rst:31]: Can you use an actual empty string here?

File: velox/docs/monitoring/stats.rst [hunk 21]
```diff
@@ -18,24 +18,50 @@ analysis.
...
      - The number of times that the memory arbitration to reclaim memory from
        an spillable operator.
        This stats only applies for spillable operators.
    * - memoryReclaimWallNanos
-     - kNano
+     - kNanos
```
Comment (mbasmanova) [inline velox/docs/monitoring/stats.rst:36]: kNanos, kBytes are not English words. Let's use "nanos" and "bytes" without quotes.

**Review by mbasmanova — COMMENTED** [review]
@duanmeng Thank you. Looks great % a couple remaining comments.

File: velox/docs/monitoring/stats.rst (hunk 21 above)
Comment (duanmeng) [reply velox/docs/monitoring/stats.rst]: Fixed

**Review by mbasmanova — APPROVED** [review]
@duanmeng Thank you for iterating. Looks great to me. Please, double check that docs render nicely.

**Comment by duanmeng** [comment]
> @duanmeng Thank you for iterating. Looks great to me. Please, double check that docs render nicely.

@mbasmanova Hi Masha, I have used the `restview` command line tool and the Clion plugin to check the render result.

**Comment by mbasmanova** [comment]
@duanmeng Would you use these instructions instead?

https://github.com/facebookincubator/velox/tree/main/velox/docs#readme

**Comment by duanmeng** [comment]
> @duanmeng Would you use these instructions instead?
> 
> https://github.com/facebookincubator/velox/tree/main/velox/docs#readme

@mbasmanova Sure, it looks like as the follow,
image

**Comment by mbasmanova** [comment]
@duanmeng This doesn't look quite right. Notice how table in HashBuild,... section is not aligned with the table in Memory Arbitration section.

**Comment by mbasmanova** [comment]
@duanmeng Thanks. This looks nice.

**Review by Yuhta — APPROVED** [review]

**Pull request closed without merging**